design/authentication-flow-usability.md

```diff
@@ -0,0 +1,322 @@
+
+5. Currently we remove whole root authentication session after authentication is finished.We can potentially change this behaviour and 
+remove just the auth session corresponding to current tab_id and remove root auth session just if there are no more auth sessions (browser tabs).
+This may fix lots of the `You are already logged-in` cases as auth session will be available in the root auth session.
+However it will have impact on memory and at the same time, there will be still cases when the root
+authentication session is expired and hence redirect to the application will be needed.
+
+
+#### Minimal approach - challenges
+
+1. From the user point of view, when the username/password screen is shown in tab2 in step 5 from [above](#the-you-are-already-logged-in-issue),
+the user will be just automatically redirected as authenticated to the application after confirm username/password. Few browser redirects is "implementation detail", which most of the users won't notice.
+However there could be some usability confusions. For example:
+
+    * User will be authenticated even if he provides invalid username/password or even he clicks "Submit" without providing any username/password. As username/password verification didn't really happen in tab2, even if it looks to user as it happened. IMO this is not major issue and same behaviour can be seen by Twitter for example.
+    
+    * After confirm username/password screen, user won't see TOTP (or any other 2 factor authenticators) even if he potentially has 2 factor enabled. He will be directly redirected to the application. I am not sure about usability of this and if it's confusing to users or if it's ok.
+
+2. Another potential challenge is handling of the error on the adapter side.
+
+    * In OAuth2/OIDC, there doesn't seem to be any nice error code for this purpose. See https://openid.net/specs/openid-connect-core-1_0.html#AuthError and https://tools.ietf.org/html/rfc6749#section-4.1.2. It seems best error code might be "server_error" and we may add the custom error_description, so that our adapters can understand it and handle it.
+
+3. In our adapters, we may need to doublecheck if behaviour is correct and user will be redirected back to KC as described in
+the step 3 [above](#minimal-approach). For servlet application, I suppose it won't be an issue as redirect_uri is usually "secured" URI and hence automatically intercepted.
+I think it may be similar for other server adapters (node.js, gatekeeper etc) but will be good to doublecheck.
+For the keycloak.js, I think "login_required" will work fine and automatically redirect to KC, but for the applications with "check_sso" or without any "onLoad" value used,
+(For example _js-console_ application from the deprecated keycloak-examples), the adapter won't automatically redirect to Keycloak. So some changes in the adapter
+may be needed to achieve this.
```

Comment:
Is checking adapters really needed? For the minimal approach, the worst what can happen is that user would have to click onto "login" button in the application which IMO is acceptable for the minimal solution.

Reply:
I was thinking about that before and IMHO the usability of that approach would be even worse than the usability of the current approach with showing the "You are already logged-in" page.

Currently when user is authenticated in tab1 and he fills the username/password in tab2, he is redirected to the "You are already logged-in" page. This is not ideal, but at least, user knows that something strange is going on. So he clicks on the "Back to application" and he would notice that he is not authenticated as he is aware of the strange/non-standard behaviour due the "You are already logged-in" page previously shown. 

With the approach you mention, after fill username/password screen, he will be redirected to the application like "js-console", but still in the anonymous mode. This can be quite confusing and user may think like "Why I am still not authenticated even if I just filled username/password" ? It can even have some security applications in theory (EG. user may not notice that he is not yet authenticated in the application and he may start to fill some forms in the application when still being in the anonymous mode etc).

That's the reason I propose to add some protocol-specific error parameters to the redirectUri to give adapters a chance to react on them and automatically redirect user back to the KC, which will do SSO dance and redirect user back as properly authenticated.